File: rust/load_balancer/src/round_robin_load_balancer.rs
use crate::backend::Backend;
use crate::health::Health;
use crate::internal_error::InternalError;
use crate::load_balancer::LoadBalancer;

use async_trait::async_trait;
use log::{debug, info};
use tokio::sync::RwLock as TokioRwLock;

/// Represents a very basic load balancer. Sends the requests to healthy backend servers in a round
/// robin fashion.
#[derive(Debug)]
pub struct RoundRobinLoadBalancer {
    /// List of backend servers
    backends: Vec<Box<dyn Backend>>,

    /// Index of the current backend server to which the next request will be sent.
    current_backend_index: TokioRwLock<usize>,
}

impl RoundRobinLoadBalancer {
    /// Creates a new load balancer with the given list of backend servers to route the requests
    /// to. The health check interval is the time in seconds between each health check sent to the
    /// backends.
    pub fn new(backends: Vec<Box<dyn Backend>>) -> Self {
        Self {
            backends,
            current_backend_index: 0.into(),
        }
    }
}

#[async_trait]
impl LoadBalancer for RoundRobinLoadBalancer {
    /// Returns the next available backend server to which the request can be sent. If none are
    /// available, an error is returned.
    async fn next_available_backend(&self) -> Result<Box<dyn Backend>, String> {
        debug!("trying to acquire current_backend_index write lock");
        let mut current_backend_index = self.current_backend_index.write().await;
        debug!("acquired current_backend_index write lock");

        let mut tried_backends = 0;

        let mut backend_index = *current_backend_index;
        *current_backend_index = (*current_backend_index + 1) % self.backends.len();

        self.backends[backend_index].check_health().await;
        let mut backend_health = self.backends[backend_index].health().await;

        while tried_backends < self.backends.len() {
            if backend_health == Health::Healthy {
                debug!("selected healthy backend {:?}", backend_index);
                return Ok(self.backends[backend_index].clone());
            }

            backend_index = *current_backend_index;
            *current_backend_index = (*current_backend_index + 1) % self.backends.len();

            self.backends[backend_index].check_health().await;
            backend_health = self.backends[backend_index].health().await;

            tried_backends += 1;
        }

        return Err("No backend server available".to_string());
    }

    /// Sends a request to the next available backend server. Returns an error if no backend server
    /// is reachable.
    async fn send_request(&self) -> Result<String, InternalError> {
        debug!("trying to get next available backend");
        let backend = self.next_available_backend().await;
        match backend {
            Ok(backend) => {
                info!("Sending request to backend {:?}", backend);
                let response = backend.send_request().await;
                match response {
                    Ok(response) => {
                        info!("{:?}", response);
                        let body = response.text_with_charset("utf-8").await.unwrap();
                        Ok(body)
                    }
                    Err(_) => Err(InternalError::BackendUnreachable),
                }
            }
            Err(_) => Err(InternalError::NoBackendAvailable),
        }
    }

    /// Checks and update the health status of all backend servers.
    async fn check_backends_healths(&self) {
        // This is used for profiling only
        let start_time = std::time::Instant::now();

        for backend in &self.backends {
            backend.check_health().await;
        }

        // For profiling only, measures how much time it took to check all backends health
        let end_time = std::time::Instant::now();
        let elapsed_time = end_time.duration_since(start_time).as_millis();
        info!("checking all backends health took {}ms", elapsed_time);
    }
}

File: cpp/source/round_robin_load_balancer.cpp
#include <round_robin_load_balancer.h>
#include <spdlog/spdlog.h>

namespace load_balancer {

RoundRobinLoadBalancer::RoundRobinLoadBalancer(std::vector<std::shared_ptr<Backend>> backends,
                                               int health_check_interval_s)
    : backends{backends},
      health_check_interval_s{health_check_interval_s},
      current_backend_index{0},
      health_check_thread_running{false},
      backend_semaphore{1} {
  spdlog::info("Load balancer started with {} backends", backends.size());
  for (auto &backend : backends) {
    spdlog::info("Backend at {} with response time {}", backend->address(),
                 backend->response_time().count());
  }
}

std::shared_ptr<Backend> RoundRobinLoadBalancer::next_available_backend() {
  int tried_backends = 0;

  this->backend_semaphore.acquire();
  while (this->backends[this->current_backend_index]->health() == Health::Unhealthy) {
    spdlog::info("Skipping unhealthy backend at {} with index {}",
                 this->backends[this->current_backend_index]->address(),
                 this->current_backend_index.load());
    this->current_backend_index = (this->current_backend_index + 1) % this->backends.size();

    tried_backends++;
    if (tried_backends >= (int)this->backends.size()) {
      spdlog::error("No healthy backends out of {} available", this->backends.size());
      this->backend_semaphore.release();
      throw std::runtime_error("No healthy backends available");
    }
  }

  auto index_to_return = this->current_backend_index.load();

  spdlog::info("Returning backend at {}", this->backends[index_to_return]->address());

  this->current_backend_index = (this->current_backend_index + 1) % this->backends.size();

  auto backend = this->backends[index_to_return];
  this->backend_semaphore.release();

  return backend;
}

drogon::Task<void> RoundRobinLoadBalancer::check_backend_healths() {
  spdlog::info("Checking health of all backends");

  auto start = std::chrono::high_resolution_clock::now();

  for (auto &backend : backends) {
    co_await backend->check_health();
  }

  auto end = std::chrono::high_resolution_clock::now();
  auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
  spdlog::info("Health check of all backends took {}ms", duration.count());
  co_return;
}

void RoundRobinLoadBalancer::start_health_checks() {
  this->health_check_thread_running = true;
  spdlog::info("Starting health checks every {}s", this->health_check_interval_s);

  drogon::async_run([this]() -> drogon::Task<void> {
    while (this->health_check_thread_running) {
      try {
        co_await this->check_backend_healths();
      } catch (const std::exception &e) {
        spdlog::error("Error checking backend healths: {}", e.what());
      }
      co_await drogon::sleepCoro(trantor::EventLoop::getEventLoopOfCurrentThread(),
                                 std::chrono::seconds(this->health_check_interval_s));
    }
    spdlog::info("Stopped health checks");
    co_return;
  });
}

void RoundRobinLoadBalancer::stop_health_checks() {
  spdlog::info("Stopping health checks");
  this->health_check_thread_running = false;
}

drogon::Task<drogon::HttpResponsePtr> RoundRobinLoadBalancer::send_request(
    drogon::HttpRequestPtr request) {
  spdlog::info("Trying to get next available backend");
  auto backend = this->next_available_backend();
  spdlog::info("Got next available backend");

  spdlog::info("Sending request to backend at {}", backend->address());
  auto response = co_await backend->send_request(request);
  spdlog::info("Received response from backend at {}", backend->address());
  co_return response;
}

} /* namespace load_balancer */
